### ISSUE TITLE
Training slows down dramatically as epochs proceed in CPU

### ISSUE THREAD

**Issue by soramichi** [chainer/chainer#2554]
The speed of a training dramatically slows down as epochs proceed in the CPU mode, especially when the batch size is small. In the example below, **the 1st epoch took 30 seconds while the 3rd one took 39 seconds**. 

The same was observed with multiple CPU models (Core i5 6400, Xeon E5-2699v3), multiple Python versions (2.7.13, 3.6.0), and multiple chainer versions (1.23, 2.0.0b1).

```
~/src/chainer/examples/mnist $ export OMP_NUM_THREADS=1
~/src/chainer/examples/mnist $ ./train_mnist.py -g -1 -b 50 -e 3
GPU: -1
# unit: 1000
# Minibatch-size: 50
# epoch: 5

epoch       main/loss   validation/main/loss  main/accuracy  validation/main/accuracy  elapsed_time
1           0.179801    0.108207              0.94595        0.9655                    30.1475       
2           0.0772594   0.0726388             0.976          0.9757                    67.1415       
3           0.0519136   0.0821774             0.983467       0.9756                    106.503 
```

The OS is Debian 9 and gcc is 6.3 (which might be too new). Python runtime and libraries were installed via Anaconda. `export OMP_NUM_THREADS=1` limits the number of threads invoked by BLAS to 1 (I use this because using more than 1 thread does not improve the performance as the matrices processed are too small in mnist).

The below is profiling results of `update_core_cpu()` in `optimizers/adam.py` using [line_profiler](https://github.com/rkern/line_profiler). The `Time` column shows the elapsed time for each line in micro seconds, which shows that **in the 3rd epoch, lines 60 and 62 took almost twice as long time as in the 1st epoch**.

- 1st epoch
```
Total time: 14.883 s
File: /home/soramichi/src/anaconda2/lib/python2.7/site-packages/chainer/optimizers/adam.py
Function: update_core_cpu at line 52

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
    52                                               @profile
    53                                               def update_core_cpu(self, param):
    54      7200        14427      2.0      0.1          grad = param.grad
    55      7200         4218      0.6      0.0          if grad is None:
    56                                                       return
    57      7200         3815      0.5      0.0          hp = self.hyperparam
    58      7200        11194      1.6      0.1          m, v = self.state['m'], self.state['v']
    59                                           
    60      7200      4065881    564.7     27.3          m += (1 - hp.beta1) * (grad - m)
    61      7200      3986827    553.7     26.8          v += (1 - hp.beta2) * (grad * grad - v)
    62      7200      6796601    944.0     45.7          param.data -= self.lr * m / (numpy.sqrt(v) + hp.eps)
```

- 3rd epoch
```
Total time: 22.6429 s
File: /home/soramichi/src/anaconda2/lib/python2.7/site-packages/chainer/optimizers/adam.py
Function: update_core_cpu at line 52

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
    52                                               @profile
    53                                               def update_core_cpu(self, param):
    54      7200        14359      2.0      0.1          grad = param.grad
    55      7200         4162      0.6      0.0          if grad is None:
    56                                                       return
    57      7200         3933      0.5      0.0          hp = self.hyperparam
    58      7200        11482      1.6      0.1          m, v = self.state['m'], self.state['v']
    59                                           
    60      7200      7869889   1093.0     34.8          m += (1 - hp.beta1) * (grad - m)
    61      7200      3928259    545.6     17.3          v += (1 - hp.beta2) * (grad * grad - v)
    62      7200     10810785   1501.5     47.7          param.data -= self.lr * m / (numpy.sqrt(v) + hp.eps)
```

---
How to reproduce the profiling results:

1. Run the mnist example and save a snapshot at the end of every epoch: ` ./train_mnist.py -g -1 -b 50 -e 3 -f 1`.
2. Put `@profile` decorator to `update_cpu_core` function of `optimizers/adam.py`.
3. In order to take a profile only for the target epoch (epoch 1 or 3), add an extension, to the mnist example, that raises an exception at the end of the target epoch.
```Python
# train_mnist.py
...
def die(trainer):
    raise ValueError("die")
...
def main():
    ...
    # die after the 1st epoch 
    trainer.extend(die, trigger=(1, 'epoch'))
    ...
```
4. Resume the mnist example from the target epoch, using `kernprof` command (usable after `pip install line_profiler`).
```
# dies after the 1st epoch finishes,
# so that the profiling result contains only the 1st epoch
$ kernprof -v -l ./train_mnist.py -g -1 -b 50 -e 3 

# resume from the end of 2nd epoch, dies when the 3rd epoch finishes
# (don't forget to change the trigger of the die extension)
$  kernprof -v -l ./train_mnist.py -g -1 -b 50 -e 3 --resume result/snapshot_iter_2400
```

**Comment by niboshi** [comment]
Thank you very much for reporting the issue!
It has reproduced in my environment.
It's still under investigation, but I'd like to share more detail of the situation.

I divided the computation in `update_core_cpu()` into more lines and ran line profiler.
(The result of all lines is at the end of this post.)

Below is excerpt from the result:
```
            Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
1st epoch:      67      3600       551386    153.2      7.1          t3 = t1 * t2
10th epoch:     67      3600      1980032    550.0     18.3          t3 = t1 * t2

1st epoch:      76      3600       579130    160.9      7.4          t4 = t1 * t3
10th epoch:     76      3600       570233    158.4      5.3          t4 = t1 * t3

1st epoch:      82      3600       676824    188.0      8.7          t1 = lr * m
10th epoch:     82      3600      2057452    571.5     19.0          t1 = lr * m

1st epoch:      84      3600       652750    181.3      8.4          t3 = t1 / t2
10th epoch:     84      3600       990012    275.0      9.2          t3 = t1 / t2
```
As can be seen from above, two multiplication (scalar * tensor) operation at line 67 and 82 is mostly responsible for the slow down. Division (matrix / tensor) at 84 also slowed down, but no so much.
I wonder why multiplication at line 76 did not slow down at alll (also scalar * tensor)...

1st epoch:
```
Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
    53                                               def update_core_cpu(self, param):
    54      3600         7329      2.0      0.1          grad = param.grad
    55      3600         2095      0.6      0.0          if grad is None:
    56                                                       return
    57      3600         2003      0.6      0.0          hp = self.hyperparam
    58      3600         4937      1.4      0.1          m, v = self.state['m'], self.state['v']
    59                                           
    60      3600        17336      4.8      0.2          beta1 = hp.beta1
    61      3600         5532      1.5      0.1          beta2 = hp.beta2
    62      3600         5843      1.6      0.1          eps = hp.eps
    63                                           
    64                                                   global count
    65      3600         3466      1.0      0.0          t1 = (1 - beta1)
    66      3600       851747    236.6     10.9          t2 = (grad - m)
    67      3600       551386    153.2      7.1          t3 = t1 * t2
    68      3600         7734      2.1      0.1          count += t2.size
    69      3600       509310    141.5      6.5          m += t3
    70      3600         5853      1.6      0.1          del t1, t2, t3
    71                                           
    72                                                   #m += (1 - hp.beta1) * (grad - m)
    73      3600         3180      0.9      0.0          t1 = (1 - beta2)
    74      3600       686510    190.7      8.8          t2 = (grad * grad)
    75      3600       765957    212.8      9.8          t3 = t2 - v
    76      3600       579130    160.9      7.4          t4 = t1 * t3
    77      3600       514649    143.0      6.6          v += t4
    78                                                   #v += (1 - beta2) * (grad * grad - v)
    79      3600         5311      1.5      0.1          del t1, t2, t3, t4
    80                                           
    81      3600        69732     19.4      0.9          lr = self.lr
    82      3600       676824    188.0      8.7          t1 = lr * m
    83      3600      1184428    329.0     15.2          t2 = (numpy.sqrt(v) + eps)
    84      3600       652750    181.3      8.4          t3 = t1 / t2
    85      3600       678157    188.4      8.7          param.data -= t3
    86      3600         5743      1.6      0.1          del t1, t2, t3
```

10th epoch:
```
Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
    53                                               def update_core_cpu(self, param):
    54      3600         7037      2.0      0.1          grad = param.grad
    55      3600         2180      0.6      0.0          if grad is None:
    56                                                       return
    57      3600         1984      0.6      0.0          hp = self.hyperparam
    58      3600         4895      1.4      0.0          m, v = self.state['m'], self.state['v']
    59                                           
    60      3600        16885      4.7      0.2          beta1 = hp.beta1
    61      3600         5588      1.6      0.1          beta2 = hp.beta2
    62      3600         6001      1.7      0.1          eps = hp.eps
    63                                           
    64                                                   global count
    65      3600         3495      1.0      0.0          t1 = (1 - beta1)
    66      3600       796798    221.3      7.4          t2 = (grad - m)
    67      3600      1980032    550.0     18.3          t3 = t1 * t2
    68      3600         7005      1.9      0.1          count += t2.size
    69      3600       498478    138.5      4.6          m += t3
    70      3600         5239      1.5      0.0          del t1, t2, t3
    71                                           
    72                                                   #m += (1 - hp.beta1) * (grad - m)
    73      3600         3170      0.9      0.0          t1 = (1 - beta2)
    74      3600       749467    208.2      6.9          t2 = (grad * grad)
    75      3600       758018    210.6      7.0          t3 = t2 - v
    76      3600       570233    158.4      5.3          t4 = t1 * t3
    77      3600       505351    140.4      4.7          v += t4
    78                                                   #v += (1 - beta2) * (grad * grad - v)
    79      3600         5294      1.5      0.0          del t1, t2, t3, t4
    80                                           
    81      3600        68229     19.0      0.6          lr = self.lr
    82      3600      2057452    571.5     19.0          t1 = lr * m
    83      3600      1172954    325.8     10.9          t2 = (numpy.sqrt(v) + eps)
    84      3600       990012    275.0      9.2          t3 = t1 / t2
    85      3600       587421    163.2      5.4          param.data -= t3
    86      3600         5275      1.5      0.0          del t1, t2, t3
```

**Comment by soramichi** [comment]
@niboshi Thank you for the investigation.

I suspect [this post](http://stackoverflow.com/questions/16553660/why-does-iterative-elementwise-array-multiplication-slow-down-in-numpy) might be relevant. What it says is numpy gets slower when the operands are super small (called [denormal](https://en.wikipedia.org/wiki/Denormal_number)), which require special handling because they do fit in the normal format of floating point numbers.

**Comment by niboshi** [comment]
@soramichi 
Thank you for helpful information!
I confirmed that adding tiny number to tensors before multiplication improves the performance.
So it's almost certain that denormal numbers are responsible for the issue, as you suggested.
We're considering how to resolve it.

**Comment by soramichi** [comment]
@niboshi 

Hi, how is this thing going on?

I guess the right way to go is "won't fix" because it only happens when some of the parameter values are extremely close to zero. This means that the NN is extremely confident of its prediction, but I don't think it happens so often in real large networks. Also, it seems to me that the only way to fix it is to add a tiny value to the almost-zero variables, which affects the learning result thus this way is not easily adopted.

However it is good to know how the core developers thought about this issue and concluded, and that information can be beneficial for similar issues that may or may not appear in the future or in similar frameworks.

Thanks.

**Comment by okuta** [comment]
How about setting these CPU flags?
https://software.intel.com/en-us/articles/x87-and-sse-floating-point-assists-in-ia-32-flush-to-zero-ftz-and-denormals-are-zero-daz

Maybe, you can use Cython.

**Comment by okuta** [comment]
I got following result on my patched branch.
```
-[7473]% ./train_mnist.py -g -1 -b 50 -e 3
GPU: -1
# unit: 1000
# Minibatch-size: 50
# epoch: 3

epoch       main/loss   validation/main/loss  main/accuracy  validation/main/accuracy  elapsed_time
1           0.183173    0.0848753             0.943833       0.9744                    31.0003
2           0.0797964   0.0882871             0.974933       0.9723                    61.4667
3           0.0537355   0.0915037             0.983383       0.9707                    91.8465
./train_mnist.py -g -1 -b 50 -e 3  94.00s user 0.37s system 100% cpu 1:34.31 total
```

Please check [this branch](https://github.com/chainer/chainer/compare/master...okuta:use-daz-ftz).
This is very simple patch.

**Comment by soramichi** [comment]
Thanks for the patch and the link. It perfectly works in my environment too and the training speed no longer degrades (still got a good accuracy).

Now the reason of this issue is clear and a solution is given, so I will close this issue.

Do you have any plan to enable these CPU flags in the upstream? Maybe giving users a choice on if the the flags are enabled is a reasonable thing to do (rather than automatically enabling them)?

**Comment by niboshi** [comment]
I reopen the issue. It should be closed only after the fix is merged to the code.

**Comment by MahdiKalayeh** [comment]
Can we use [this patch](https://github.com/chainer/chainer/compare/master...okuta:use-daz-ftz) for GPU as well? If not, would you please provide one. I am observing the same issue while running on GPU. I should note that I am using latest Cupy.

**Comment by soramichi** [comment]
@MahdiKalayeh 

No, `_mm_setcs` is to set a control register of SIMD instructions equipped in CPUs.

Also, as I don't observe the same phenomenon when I run the same code in a GPU,
it will be helpful if you could report what you did and what you observed in detail.

**Comment by MahdiKalayeh** [comment]
@soramichi Below you can see the log of my code while training

train rcnn after 0.10 hours: 001/0500 >>  loss:0.1925  ap:0.0420
train rcnn after 0.21 hours: 001/1000 >>  loss:0.1595  ap:0.0509
train rcnn after 0.31 hours: 001/1500 >>  loss:0.1581  ap:0.0535
train rcnn after 0.42 hours: 001/2000 >>  loss:0.1534  ap:0.0551
train rcnn after 0.53 hours: 001/2500 >>  loss:0.1550  ap:0.0619
train rcnn after 0.65 hours: 001/3000 >>  loss:0.1523  ap:0.0641
train rcnn after 0.75 hours: 001/3500 >>  loss:0.1503  ap:0.0688
train rcnn after 0.88 hours: 001/4000 >>  loss:0.1471  ap:0.0677
train rcnn after 1.03 hours: 001/4500 >>  loss:0.1495  ap:0.0813
train rcnn after 1.19 hours: 001/5000 >>  loss:0.1438  ap:0.0804
train rcnn after 1.34 hours: 001/5500 >>  loss:0.1454  ap:0.0846
train rcnn after 1.49 hours: 001/6000 >>  loss:0.1484  ap:0.0884
train rcnn after 1.65 hours: 001/6500 >>  loss:0.1494  ap:0.0880
train rcnn after 1.82 hours: 001/7000 >>  loss:0.1454  ap:0.0908
train rcnn after 1.98 hours: 001/7500 >>  loss:0.1463  ap:0.0948
validation rcnn >>  loss:0.1957  ap:0.1162
train rcnn after 0.30 hours: 002/0500 >>  loss:0.1432  ap:0.1104
train rcnn after 0.64 hours: 002/1000 >>  loss:0.1381  ap:0.1110
train rcnn after 1.01 hours: 002/1500 >>  loss:0.1397  ap:0.1116
train rcnn after 1.37 hours: 002/2000 >>  loss:0.1444  ap:0.1125
train rcnn after 1.75 hours: 002/2500 >>  loss:0.1420  ap:0.1191
train rcnn after 2.13 hours: 002/3000 >>  loss:0.1420  ap:0.1094
train rcnn after 2.52 hours: 002/3500 >>  loss:0.1434  ap:0.1166
train rcnn after 2.90 hours: 002/4000 >>  loss:0.1413  ap:0.1219

As you can see, 500 iterations initially takes 0.1 hours while this number increases to 0.16 at the last 500 instances of the first epoch. Beginning of the second epoch, it takes 0.3 hours for 500 iterations which continuously increases to 0.48. I am using chainer 2.0.0 and cupy 1.0.0.1. Clearly, training is drastically slowing down despite I thought it should take less time due to caching of IO. Can a CPU bottleneck grow in time? I am using iterators.MultiprocessIterator but I can see that GPU utilization fluctuates.

**Comment by soramichi** [comment]
Well obviously your case includes too much factors and the problem cannot be really clear.

I recommend you to do (1) make the case simpler to dig the real problem (for example making your network smaller, using only 1 GPU, ...) and (2) profile your code after it has been made sufficiently simple.

About the denormal numbers discussed in this issue, you can see if they appear in your network by dumping the parameters from the network with [params method](https://docs.chainer.org/en/stable/reference/core/link.html#chainer.Link.params).

**Comment by MahdiKalayeh** [comment]
Update: I noticed that 1) each instance which iterator creates, is very large (~ 200 MB) as I am dealing with videos and 2) I do not use ~25% of read data (e.g I read some images from disk but never use them due to some sub-sampling).

Modifications: I simply reduced the number of frames per video (e.g do not read images that you won't use later) and now each instance is approximately 10 MB. I do not see growing execution time anymore. It is almost constant after 4 epochs and I tried it on three different machines. This is with exact same network but with a lighter iterator load and no unnecessary IO.

Next: As soon as I have some time, I will do the recommended profiling.

**Comment by soramichi** [comment]
@niboshi said they "reopen the issue. It should be closed only after the fix is merged to the code". Closing staled issues makes the status look cleaner but it actually helps nothing.

**Comment by niboshi** [comment]
Yes, this issue should be kept open.

(Please consider the stale bot as a moderate notification of forgotten issues. It's completely up to those participated whether issues are to be closed or not. Just say something and the stale label will be canceled.)

**Comment by okuta** [comment]
I made a Python package called `daz`. It has function that control CPU flags.
